tests/subsys/net/lib/nrf_cloud/src/main.c

```diff
@@ -0,0 +1,153 @@
+/*
+ * Copyright (c) 2022 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
+ */
+
+#include <unity.h>
+#include <zephyr.h>
+#include <stdio.h>
+#include <net/nrf_cloud.h>
+#include <mock_nrf_cloud_transport.h>
+#include "nrf_cloud_fsm.h"
+
+/* This is required since unity_main return int and
+ * zephyr expects main not to return any value
+ */
+extern int unity_main(void);
+
+/* The library can only be initialized when the internal state is
+ * idle, so the test verifies that nrf_cloud_init function will
+ * return with permission denied error (-EACCES) when not idle
+ */
+void test_init_not_idle(void)
+{
+	struct nrf_cloud_init_param params = {
+		.event_handler = NULL,
+		.client_id = NULL,
+	};
+
+	/* nfsm_set_current_state_and_notify directly sets the
+	 * internal state of the nrf_cloud lib, allowing us to
+	 * easily test various initialization scenarios
+	 */
+	nfsm_set_current_state_and_notify(STATE_TOTAL, NULL);
```

Comment:
I wouldn't use `STATE_TOTAL` since it is not a valid state at all

I recommend calling `nrf_cloud_init` twice instead, since the first init call will get the library into a non-idle state

Reply:
I think that `nrf_cloud_init` should not be called twice, since it is under test. There is no guarantee it will not fail when being called the first time. `STATE_IDLE` is not a valid state but it can be good enough for testing.

Reply:
> There is no guarantee it will not fail when being called the first time

I think your test case `test_init_success` covers this possibility sufficiently, and I lean towards tests that mess with the internal state as little as possible to make sure that what we are testing is actual behavior

Though if you choose to accept my suggestion, one thing I would also add are some assertions: `nfsm_get_current_state` is `STATE_IDLE` before the test runs, and an assertion that `nfsm_get_current_state` is `STATE_INITIALIZED`  after the first call.

Reply:
I agree with @glarsennordic. If you look at SMS tests (same for location library), everything is basically tested through the public API of the library plus the event handlers underneath some libraries that it's using (`at_monitor_dispatch()`). 

If possible, I'd advise always to avoid using internal functions for testing. They might be needed and required in some cases but the idea should be not to use them if you have a way of avoiding it.

Reply:
Great, I'll see what I can change to limit these internal calls.